### ISSUE TITLE
Ditch Unity in favour of MonoGame

### ISSUE THREAD

**Issue by acidiclight** [sociallydistantgame/sociallydistant#8]
Unity is a great game engine if you enjoy sadomasochism and, in some cases, financial grief. We all know about the runtime fee mess, but I'm more frustrated by the engine's UI system and many other small annoyances I have to deal with when programming Socially Distant.

## What annoys me

 - Unity's auto layout system has no concept of a "maximum size" for layout elements - leading to the need for parent elements that define the max size of their child. This is annoying as it clutters the hierarchy.
 - When opening UI prefabs, the scene view sometimes defaults to 3D view when you almost never want this.
 - When editing a UI prefab, you can't see it in Game View and it's almost impossible to get a unit-perfect preview of what it'll look like.
 - Implementing `IPointerClickHandler` means nothing unless you implement `IPointerDownHandler` and `IPointerUpHandler`, even if you don't need them, meaning you have public stubs that you now have to document as such.
 - `MonoBehaviour` and `ScriptableObject` types do not have user-defined constructors (because they are glued to C++ objects that Unity must allocate), meaning dependency injection and encapsulation is a bitch.
 - Unity struggles to deal with Nullable Reference Types, this violates my coding style.
 - Unity Hub pisses me off and keeps launching every time I run the editor under Linux
 - Speaking of Linux, we need to do a stupid workaround with a shell script just to get assets to import. This bug has been a bug for over a yea and is still a bug in Unity 6. **Why?**
 - Licensing Pro cost more, diregarding runtime fee. We may never need it, but I don't want to deal with the risk. I can't afford it.
 - Unity isn't open-ource, I cannot fix it.
 - Unity crashes every 2 nanoseconds.
 - Unless I want to accept the runtime fee term, I'm stuck with a splash screen.
 - If I were to upgrade to Pro, any contributors here would need to as well. I can't afford that. You can't afford that. (EULA for 2021 LTS bans mixing different licenses in the same project)
 - We keep getting mono crash dumps in the repo root despite me having gitignored them. Fuck.
 - There's a shit ton of annoying `.meta` files cluttering the project, slowing Rider search to a crawl, even on my NVMe drive. Fucking hell.
 - We can't use NuGet
 - We're stuck on Mono
 - TextMeshPro `<mark>` elements render over text, not behind it
 - Asset bundle building isn't automatic, I had to automate it myself. By doing so, I introduced a bug where if your code accidentally imports an editor-only lib, the Unity Editor will infinite-loop. I have to `pkill` it and figure out what script caused that.
 - Unity Editor doesn't use File Chooser Portal under Linux, and there are MANY UI/UX issues as a result of it using GTK file dialogs incorrectly. Try building a standalone game.
 - Sometimes Unity decides not to save my editor preferences.
 - Sometimes Unity decides the game is not worthy of being opened.
 - I get yelled at for not specifically using Ubuntu LTS if I try to report a crash, even though this is a desktop system with a11y needs and I **need** up-to-date packages. Yes, that means I use Arch by the way.

That's a lot.

## What MonoGame solves

MonoGame is open-source under MS-PL, and it's not a game engine. It's a game framework, meaning we do lose a lot of engine features, which actually does solve a lot of problems for us. Because it'll be us solving them.

 - All UI-related issues will be solved by writing a UI system that actually does things properly.
 - The Asset Bundle bug will be solved by writing my own tools that don't infinite-loop because of a `using` directive in a random C# script. There are no editor-only assemblies.
 - MonoGame doesn't charge me for others downloading the game.
 - MonoGame doesn't force a splash screen on me.
 - It won't crash unless I crash it.
 - MonoGame Hub doesn't exist. It's not a game engine, there are no licenses or project files.
 - There are no editor preferences for it to forget to save, there's no editor.
 - TextMeshPro `<mark>` elements are something I can program myself, correctly.
 - You don't need to work around an asset importer bug on Linux.
 - You don't need Ubuntu LTS to get support.
 - We can upstream bugfixes, if any.
 - We can ue NuGet.
 - We can use modern .NET.

## What it'd break

Most of the game is engine-agnostic. That's great, means porting is a matter of ctrl+c ctrl+v.

The user interface isn't.

The content manager isn't.

Asset Bundles would need to be recreated, MonoGame's asset system is insufficient.

We'd need to write a UI system. One that supports all the same stuff I do in Unity, without me needing to fight with Unity. This is challenging and will take weeks, if not months.

So basically....while most of the game won't need to be touched save fdor some namespace changes and minor refactoring, the UI and everything you **see** in the game will need to be rewritten. It'd almost be a development reset, and would set us back a couple devlogs worth of work.
